src/What4/ExprHelpers.hs

```diff
@@ -115,10 +117,69 @@
 import qualified What4.SemiRing as SR
 import qualified What4.Expr.BoolMap as BM
 import qualified What4.Symbol as WS
+import qualified What4.Utils.AbstractDomains as W4AD
 
 import           Data.Parameterized.SetF (SetF)
 import qualified Data.Parameterized.SetF as SetF
 
+-- | Sets the abstract domain of the given integer to assume
+--   that it is positive. 
+assumePositiveInt ::
+  W4.IsExprBuilder sym =>
+  sym ->
+  W4.SymExpr sym W4.BaseIntegerType ->
+  W4.SymExpr sym W4.BaseIntegerType
+assumePositiveInt _sym e = 
+  let
+    rng = W4AD.rangeMax (W4AD.singleRange 0) (W4.integerBounds e)
+  in W4.unsafeSetAbstractValue rng e
+
+-- | Copied from https://github.com/GaloisInc/what4/commit/de5e73170b98e1f7e9c9baa9f044e2dfd21016cb
+--   Using this definition for 'intMax' allows more abstract domain information to be used
+--   in order to statically determine the result. Additionally the resulting expression
+--   contains a more precise abstract domain based on the semantics of 'max'.
+intMax ::
+  W4.IsExprBuilder sym =>
+  sym ->
+  W4.SymInteger sym ->
+  W4.SymInteger sym ->
+  IO (W4.SymInteger sym)
+intMax sym x y =
+  do x_le_y <- W4.intLe sym x y
+     y_le_x <- W4.intLe sym y x
+     case (W4.asConstantPred x_le_y, W4.asConstantPred y_le_x) of
+       -- x <= y
+       (Just True, _) -> return y
+       -- x < y
+       (_, Just False) -> return y
+       -- y < x
+       (Just False, _) -> return x
+       -- y <= x
+       (_, Just True) -> return x
+       _ ->
+         do let rng_x = W4.integerBounds x
+            let rng_y = W4.integerBounds y
+            W4.unsafeSetAbstractValue (W4AD.rangeMax rng_x rng_y) <$>
+              W4.intIte sym x_le_y y x
+
+-- | Overriding 'W4.integerToNat' to use a modified 'intMax' that
+--   statically evaluates in more cases. In particular this ensures that
+--   integers with a positive abstract domain aren't wrapped in any additional checks,
+--   and that the inner integer is statically known to be positive.
+integerToNat ::
+  W4.IsExprBuilder sym =>
```

Comment:
I'm obviously always worried about unsafeCoerce. Is there a time horizon to replace this with safe what4 alternatives now that we have improved infrastructure there? If so, can you reference the ticket here?